text/0000-contingency-plan.md

```diff
@@ -0,0 +1,425 @@
+  currently, opt-in is limited to "surface changes" only, but allowing
+  opt-in to affect the type system feels like it would be creating two
+  distinct languages.
+- It seems likely that all users of Rust will want to know that their code
+  is sound and would not want to be working with unsafe constructs or bugs.
+- Users may choose not to opt-in to newer versions because they do not
+  need the new features introduced there or because they wish to
+  preserve compatibility with older compilers. It would be sad for
+  them to lose the benefits of bug fixes as well.
+- We already have several mitigation measures, such as opt-out or
+  temporary deprecation, that can be used to ease the transition
+  around a soundness fix. Moreover, separating out new type rules so
+  that they can be "opted into" can be very difficult and would
+  complicate the compiler internally; it would also make it harder to
+  reason about the type system as a whole.
+
+**Rather than using a version number to opt-in to minor changes, one
+might consider using the existing feature mechanism.** For example,
+one could write `#![feature(foo)]` to opt in to the feature "foo" and
+its associated keywords and type rules, rather than
+`#![rust_version="1.2.3"]`. While using minimum version numbers is
+more opaque than named features, they do offer several advantages:
+
+1. Using a version number alone makes it easy to think about what
+   version of Rust you are using as a conceptual unit, rather than
+   choosing features "a la carte".
+2. Using named features, the list of features that must be attached to
+   Rust code will grow indefinitely, presuming your crate wants to
+   stay up to date.
+3. Using a version attribute preserves a mental separation between
+   "experimental work" (feature gates) and stable, new features.
+4. Named features present a combinatoric testing problem, where we
+   should (in principle) test for all possible combinations of
+   features.
+   
+# Unresolved questions
+
+**Can (and should) we give a more precise definition for compiler bugs
+and soundness problems?** The current text is vague on what precisely
+constitutes a compiler bug and soundness change. It may be worth
+defining more precisely, though likely this would be best done as part
+of writing up a more thorough (and authoritative) Rust reference
+manual.
+
+**Should we add a mechanism for "escaping" keywords?"** We may need a
+mechanism for escaping keywords in the future. Imagine you have a
+public function named `foo`, and we add a keyword `foo`. Now, if you
+opt in to the newer version of Rust, your function declaration is
+illegal: but if you rename the function `foo`, you are making a
+breaking change for your clients, which you may not wish to do. If we
+had an escaping mechanism, you would probably still want to deprecate
+`foo` in favor of a new function `bar` (since typing `foo` would be
+awkward), but it could still exist.
```

Comment:
@nrc 

> This seems so easy to fix with a refactoring tool that it should be unnecessary.

I don't follow. Point is, if I update my compiler and rename my methods, I break my clients and force them to refactor. Maybe your point is: that's acceptable, because we can have a tool that renames functions from `match` to `match_` or something? (we may itself introduce conflicts...)